probe-rs/src/architecture/riscv/communication_interface.rs

```diff
@@ -0,0 +1,848 @@
+        let mut dtmcs = Dtmcs(0);
+
+        dtmcs.set_dmireset(true);
+
+        let Dtmcs(reg_value) = dtmcs;
+
+        let bytes = reg_value.to_le_bytes();
+
+        let jtag_interface = self
+            .probe
+            .get_interface_jtag_mut()
+            .ok_or(DebugProbeError::InterfaceNotAvailable("JTAG"))?;
+
+        jtag_interface.write_register(0x10, &bytes, 32)?;
+
+        Ok(())
+    }
+
+    fn read_idcode(&mut self) -> Result<u32, DebugProbeError> {
+        let jtag_interface = self
+            .probe
+            .get_interface_jtag_mut()
+            .ok_or(DebugProbeError::InterfaceNotAvailable("JTAG"))?;
+
+        let value = jtag_interface.read_register(0x1, 32)?;
+
+        Ok(u32::from_le_bytes((&value[..]).try_into().unwrap()))
+    }
+
+    /// Perform an access to the dmi register of the JTAG Transport module.
+    ///
+    /// Every access both writes and reads from the register, which means a value is always
+    /// returned. The `op` is checked for errors, and if it is not equal to zero, an error is returned.
+    fn dmi_register_access(
+        &mut self,
+        address: u64,
+        value: u32,
+        op: DmiOperation,
+    ) -> Result<u32, RiscvError> {
+        let register_value: u128 = ((address as u128) << 34) | ((value as u128) << 2) | op as u128;
+
+        let bytes = register_value.to_le_bytes();
+
+        let bit_size = self.abits + 34;
+
+        let jtag_interface = self
+            .probe
+            .get_interface_jtag_mut()
+            .ok_or(DebugProbeError::InterfaceNotAvailable("JTAG"))?;
+
+        let response_bytes = jtag_interface.write_register(0x11, &bytes, bit_size)?;
+
+        let response_value: u128 = response_bytes.iter().enumerate().fold(0, |acc, elem| {
+            let (byte_offset, value) = elem;
+            acc + ((*value as u128) << (8 * byte_offset))
+        });
+
+        // Verify that the transfer was ok
+
+        let op = (response_value & 0x3) as u8;
+
+        if op != 0 {
+            return Err(RiscvError::DmiTransfer(
+                DmiOperationStatus::parse(op).unwrap(),
+            ));
+        }
+
+        let value = (response_value >> 2) as u32;
+
+        Ok(value)
+    }
+
+    pub(crate) fn read_dm_register<R: DebugRegister>(&mut self) -> Result<R, RiscvError> {
+        log::debug!("Reading DM register '{}' at {:#010x}", R::NAME, R::ADDRESS);
+
+        // TODO: parameter for this
+        let retry_duration = Duration::from_secs(5);
+
+        let start_time = Instant::now();
+
+        loop {
+            // Prepare the read, answer will be read back in a second operation
+            match self.dmi_register_access(R::ADDRESS as u64, 0, DmiOperation::Read) {
+                Ok(_) => break,
+                Err(RiscvError::DmiTransfer(DmiOperationStatus::RequestInProgress)) => {
+                    // Operation still in progress, reset dmi status and try again.
+                    self.dmi_reset()?;
```

Comment:
Consider adding a backoff, and also increase the jtag idle cycles?

Reply:
I tested increasing the idle cycles manually, and did not really see an improvement.
Currently, the number of idle cycles is just set to the value recommended in the `dtmcs` register. This worked fine so far for all requests, except for the system reset. In that case, a lot of retries have to occur until the module does respond again.

I'm not sure if it makes sense to increase the idle cycles for all requests, or if it would be easier to just have a special case for the reset.

Reply:
Let`s keep it like that for now. If we find a chip which causes problems, we can revisit this.

(The GD32V chip I have is weird).

I can`t really test right now, because I destroyed my last jtag (by experimentation). I think I will get some FTDI chips, and make a backend for probe-rs.